File: backend/src/main/java/tom/projects/monkeylog/model/workout/Workout.java
package tom.projects.monkeylog.model.workout;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import tom.projects.monkeylog.model.user.UserOwned;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class Workout implements UserOwned {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private Long clonedFromId;

    private String name;

    private String note;

    private Long userId;

    @Enumerated(EnumType.STRING)
    private WorkoutType workoutType;

    private LocalDateTime startDate;

    private LocalDateTime endDate;

    @OneToMany(mappedBy = "workout", cascade =  CascadeType.ALL, orphanRemoval = true)
//    @OnDelete(action = OnDeleteAction.CASCADE)
    @OrderColumn(name = "sort_order")
    private List<ExerciseGroup> exerciseGroups = new ArrayList<>();

    @ManyToOne
    @JoinColumn(name="program_week_id")
    private ProgramWeek programWeek;

    public static Workout clone(Workout workout) {
        Workout newWorkout = new Workout();
        newWorkout.setWorkoutType(WorkoutType.TEMPLATE);
        newWorkout.setName(workout.getName());
        newWorkout.setUserId(workout.getUserId());
        newWorkout.setNote(workout.getNote());
        newWorkout.setWorkoutType(workout.getWorkoutType());
        workout.getExerciseGroups().stream().map(ExerciseGroup::clone).forEach(newWorkout::addExerciseGroup);

        return newWorkout;
    }

    public void addExerciseGroup(ExerciseGroup exerciseGroup) {
        exerciseGroups.add(exerciseGroup);
        exerciseGroup.setWorkout(this);
    }

}

File: backendkt/src/main/kotlin/tom/com/monkeylog/model/workout/Workout.kt
package tom.com.monkeylog.model.workout

import jakarta.persistence.*
import org.hibernate.annotations.UuidGenerator
import tom.com.monkeylog.model.user.UserOwned
import java.time.LocalDateTime
import java.util.*

@Entity
class Workout(
    @Id
    @GeneratedValue
    @UuidGenerator
    var id: UUID? = null,
    var clonedFromId: UUID? = null,
    var name: String,
    var note: String? = null,
    override var userId: UUID? = null,
    @Enumerated(EnumType.STRING)
    var workoutType: WorkoutType = WorkoutType.TEMPLATE,
    var startDate: LocalDateTime? = null,
    var endDate: LocalDateTime? = null,
    @OneToMany(
        mappedBy = "workout",
        cascade = [CascadeType.ALL],
        orphanRemoval = true
    )
    @OrderColumn(name = "sort_order")
    var exerciseGroups: MutableList<ExerciseGroup> = ArrayList(),
    @ManyToOne
    @JoinColumn(name = "program_week_id")
    var programWeek: ProgramWeek? = null
) : UserOwned {
    fun clone(): Workout {
        return Workout(
            name = name,
            note = note,
            userId = userId,
            workoutType = workoutType,
            startDate = startDate,
            endDate = endDate,
            programWeek = programWeek,
            exerciseGroups = exerciseGroups.map { it.clone(this) }.toMutableList()
        )
    }
}